thorlcr/master/thgraphmanager.cpp

```diff
@@ -60,10 +61,157 @@ class CJobManager : public CSimpleInterface, implements IJobManager, implements
     StringAttr          currentWuid;
     ILogMsgHandler *logHandler;
 
+    CJobMaster *getCurrentJob() { CriticalBlock b(jobCrit); return jobs.ordinality() ? &OLINK(jobs.item(0)) : NULL; }
     bool executeGraph(IConstWorkUnit &workunit, const char *graphName, const SocketEndpoint &agentEp);
     void addJob(CJobMaster &job) { CriticalBlock b(jobCrit); jobs.append(job); }
     void removeJob(CJobMaster &job) { CriticalBlock b(jobCrit); jobs.zap(job); }
 
+    class CThorDebugListener : public CSimpleInterfaceOf<IInterface>, implements IThreaded
+    {
+    protected:
+        CThreaded threaded;
+        unsigned port;
+        Owned<ISocket> sock;
+        CJobManager &mgr;
+    private:
+        volatile bool running;
+    public:
+        CThorDebugListener(CJobManager &_mgr) : threaded("CThorDebugListener", this), mgr(_mgr)
+        {
+            port = globals->getPropInt("DebugPort", 9877);
+            running = true;
+            threaded.start();
+        }
+        ~CThorDebugListener()
+        {
+            running = false;
+            if (sock)
+                sock->cancel_accept();
+            threaded.join();
+        }
+        virtual unsigned getPort() { return port; }
+
+        virtual void processDebugCommand(CSafeSocket &ssock, StringBuffer &rawText)
+        {
+            Owned<IPropertyTree> queryXml;
+            try
+            {
+                queryXml.setown(createPTreeFromXMLString(rawText.str(), ipt_caseInsensitive, (PTreeReaderOptions)(ptr_ignoreWhiteSpace|ptr_ignoreNameSpaces)));
+            }
+            catch (IException *E)
+            {
+                StringBuffer s;
+                DBGLOG("ERROR: Invalid XML received from %s:%s", E->errorMessage(s).str(), rawText.str());
+                throw;
+            }
+
+            Linked<CJobMaster> job = mgr.getCurrentJob();
+            if (!job)
+                throw MakeStringException(5300, "Command not available when no graph active");//TODO:Update error code
+
+            const char *command = queryXml->queryName();
+            const char *edgeId = queryXml->queryProp("@edgeId");
+            const char *graphId = job->queryGraphName();
```

Comment:
Is this right?
graphId vs graphName .. graphName will be something like graph1. A graphId is a numeric that refers to a specific subgraph.

The graphId here isn't being used except as a header in the response. Shouldn't it be passed to the slaves?

@richardkchapman - Does debugger in Roxie/HThor send in a graphId to be debugged?